assets/content/cookbook/Data structures/ring-array.md

```diff
@@ -5,82 +5,78 @@
 A fixed ring array are especially useful when you need a hard upper bound for how much data can be in the queue.
 
 ```haxe
-@:generic class Ring<T> {
-  public var cap(get, never): Int;
-  inline function get_cap() return a.length;
+// reference https://github.com/torvalds/linux/blob/master/include/linux/circ_buf.h
+@:generic
+class Ring<T> {
 
-  public var len(get, never): Int;
-  inline function get_len() return i + left - start;
-
-  var i: Int;
-  var start: Int;
-  var left: Int;
+  var head: Int;
+  var tail: Int;
+  var cap: Int;
   var a: haxe.ds.Vector<T>;
 
+  // NOTE: len must be power of 2
   public function new(len) {
     a = new haxe.ds.Vector<T>(len);
+    cap = len - 1;
     reset();
   }
-  public function pop(): Null<T> {
-    if (len <= 0) return null;
-    if (i == 0) {
-      i = cap;
-      left = 0;
-    }
-    return a[--i];
+
+  public function reset() {
+    head = 0;
+    tail = 0;
   }
+
+  public function push(v: T) {
+    if (CIRC_SPACE() == 0) tail = (tail + 1) & cap; // override
+    a[head] = v;
+    head = (head + 1) & cap;
+  }
+
   public function shift(): Null<T> {
-    if (len <= 0) return null;
-    if (start == cap) {
-      start = 0;
-      left = 0;
+    var ret:Null<T> = null;
+    if (CIRC_CNT() > 0) {
+      ret = a[tail];
+      tail = (tail + 1) & cap;
     }
-    return a[start++];
+    return ret;
   }
-  public function push(v: T) {
-    if (i == cap) {
-      if (left > 0 && start == i) start = 0;
-      i = 0;
-      left = cap;
+
+  public function pop(): Null<T> {
+    var ret:Null<T> = null;
+    if (CIRC_CNT() > 0) {
+      head = (head - 1) & cap;
+      ret = a[head];
     }
-    if (len == cap) start++;
-    a[i++] = v;
+    return ret;
   }
-  public function reset() {
-    i = 0;
-    start = 0;
-    left = 0;
+
+  public function unshift(v: T) {
+    if (CIRC_SPACE() == 0) head = (head - 1) & cap;
+    tail = (tail - 1) & cap;
+    a[tail] = v;
   }
-  public function remove(v: T) {
-    var cap = this.cap;
-    var max = this.len;
-    var j = 0, p = 0;
-    while (j < max) {
-      p = (j + start) % cap;
-      if (v == a[p]) {
-        if (p == start) {
-          ++ start;
-        } else {
-          if (this.i == 0) {
-            this.i = cap;
-            this.left = 0;
-          }
-          -- max;
-          while (j < max) {
-            a[(j + start) % cap] = a[(j + start + 1) % cap];
-            ++ j;
-          }
-          -- this.i;
-        }
-        break;
-      }
-      ++ j;
-    }
+
+  public function toString() {
+    return '[head: $head, tail: $tail, capacity: $cap]';
   }
-  public inline function toString() {
-    return '[i: $i, start: $start, len: $len, left: $left]';
+
+  public inline function CIRC_CNT() return (head - tail) & cap;
```

Comment:
I understand the `CIRC_*` methods are taken from Linux, but you should change their names. At least `count` (`CIRC_CNT`) and `space` (`CIRC_SPACE`) can be properties with an inline getter.